File: backend/src/models/PaymentLink.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IUser } from './User';

export interface IPaymentLink extends Document {
  ownerId: IUser['_id'];
  title: string;
  description?: string;
  amount: number;
  currency: string;
  status: 'ACTIVE' | 'EXPIRED' | 'USED';
  expiresAt?: Date;
  maxUses?: number;
  currentUses: number;
  customerPhone?: string;
  customerEmail?: string;
  redirectUrl?: string;
  webhookUrl?: string;
  paymentLink: string;
  transactions: IUser['_id'][];
  createdAt: Date;
  updatedAt: Date;
}

const PaymentLinkSchema: Schema = new Schema({
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  title: { type: String, required: true },
  description: { type: String },
  amount: { type: Number, required: true },
  currency: { type: String, default: 'KES' },
  status: { type: String, enum: ['ACTIVE', 'EXPIRED', 'USED'], default: 'ACTIVE' },
  expiresAt: { type: Date },
  maxUses: { type: Number },
  currentUses: { type: Number, default: 0 },
  customerPhone: { type: String },
  customerEmail: { type: String },
  redirectUrl: { type: String },
  webhookUrl: { type: String },
  paymentLink: { type: String, required: true, unique: true },
  transactions: [{ type: Schema.Types.ObjectId, ref: 'GatewayTransaction' }],
}, { timestamps: true });

PaymentLinkSchema.index({ ownerId: 1 });
PaymentLinkSchema.index({ paymentLink: 1 });
PaymentLinkSchema.index({ status: 1 });

export default mongoose.model<IPaymentLink>('PaymentLink', PaymentLinkSchema);

File: backend/app/models/payment_link.py
from datetime import datetime, timezone
from typing import Optional, Literal
from beanie import PydanticObjectId
from pydantic import Field
from app.models.base import BaseDocument


class PaymentLink(BaseDocument):
    owner_id: PydanticObjectId = Field(alias="ownerId")
    title: str
    description: Optional[str] = None
    amount: float
    currency: str = "KES"
    status: Literal["ACTIVE", "EXPIRED", "USED"] = "ACTIVE"
    expires_at: Optional[datetime] = Field(default=None, alias="expiresAt")
    max_uses: Optional[int] = Field(default=None, alias="maxUses")
    current_uses: int = Field(default=0, alias="currentUses")
    customer_phone: Optional[str] = Field(default=None, alias="customerPhone")
    customer_email: Optional[str] = Field(default=None, alias="customerEmail")
    redirect_url: Optional[str] = Field(default=None, alias="redirectUrl")
    webhook_url: Optional[str] = Field(default=None, alias="webhookUrl")
    payment_link: str = Field(unique=True, alias="paymentLink")
    transactions: list[PydanticObjectId] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "paymentlinks"
        indexes = [
            "owner_id",
            "payment_link",
            "status",
        ]
